CONTRIBUTING.md

```diff
@@ -1,12 +1,205 @@
+- [Create your own fork](https://docs.github.com/en/get-started/quickstart/fork-a-repo) of [this repository](https://github.com/cloudflare/wrangler2).
+- Clone your fork to your local machine
+  ```sh
+  > git clone https://github.com/<your-github-username>/wrangler2
+  > cd wrangler
+  ```
+  You can see that your fork is setup as the `origin` remote repository.
+  Any changes you wish to make should be in a local branch that is then pushed to this origin remote.
+  ```sh
+  > git remote -v
+  origin	https://github.com/<your-github-username>/wrangler2 (fetch)
+  origin	https://github.com/<your-github-username>/wrangler2 (push)
+  ```
+- Add `cloudflare/wrangler2` as the `upstream` remote repository.
+  ```sh
+  > git remote add upstream https://github.com/cloudflare/wrangler2
+  > git remote -v
+  origin	https://github.com/<your-github-username>/wrangler2 (fetch)
+  origin	https://github.com/<your-github-username>/wrangler2 (push)
+  upstream	https://github.com/cloudflare/wrangler2 (fetch)
+  upstream	https://github.com/cloudflare/wrangler2 (push)
+  ```
+- You should regularly pull from the `main` branch of the `upstream` repository to keep up to date with the latest changes to the project.
+  ```sh
+  > git switch main
+  > git pull upstream main
+  From https://github.com/cloudflare/wrangler2
+  * branch            main       -> FETCH_HEAD
+  Already up to date.
+  ```
+
+### Install dependencies
+
+The Node.js dependencies of the project are managed by the [`npm`](https://www.npmjs.com/) tool.
+
+This repository is setup as a [mono-repo](https://docs.npmjs.com/cli/v7/using-npm/workspaces) of workspaces. The workspaces are stored in the [`packages`](https://github.com/cloudflare/wrangler2/tree/main/packages) directory.
+
+While each workspace has its own dependencies, you install the dependencies using `npm` at the root of the project.
+
+- Install all the dependencies
+  ```sh
+  > cd wrangler2
+  > npm install
+  ```
+
+**Do not run `npm install` in any of the workspaces directly.**
+
+## Building and running
+
+Each wrangler workspace in this project is written in [TypeScript](https://www.typescriptlang.org/) and compiled, by [esbuild](https://github.com/evanw/esbuild), into JavaScript bundles for distribution.
+
+- Run a distributable for a specific workspace (e.g. wrangler)
+  ```sh
+  > npm start -w wrangler
+  ```
+- Build a distributable for a specific workspace(e.g. wrangler)
+  ```sh
+  > npm run build -w wrangler
+  ```
+
+## Checking the code
+
+The code in the repository is checked for type checking, formatting, linting and testing errors.
+
+- Run all checks in all the workspaces
+  ```sh
+  > npm run check
+  ```
+
+When doing normal development you may want to run these checks individually.
+
+### Type Checking
+
+The code is checked for type errors by [TypeScript](https://www.typescriptlang.org/).
+
+- Type check all the code in the repository
+  ```sh
+  > npm run type-check
```

Comment:
This doesn't exist. Should we replace with `npx tsc`, or actually make a `type-check` script? Either works for me.